core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -297,21 +297,24 @@ public long getDefaultBlockSize() {
         List<WorkerNetAddress> locations = fileBlockInfo.getBlockInfo().getLocations()
             .stream().map(alluxio.wire.BlockLocation::getWorkerAddress).collect(toList());
         if (locations.isEmpty()) {
-          // No in-Alluxio location, fallback to use under file system locations with
-          // co-located workers.
+          // No in-Alluxio location
           if (workerHosts == null) {
             // lazy initialization for rpc call
             workerHosts = getHostToWorkerMap();
           }
-          Map<String, WorkerNetAddress> finalWorkerHosts = workerHosts;
-          locations = fileBlockInfo.getUfsLocations().stream()
-              .map(location -> finalWorkerHosts.get(HostAndPort.fromString(location).getHostText()))
-              .filter(Objects::nonNull).collect(toList());
-        }
-        if (locations.isEmpty()) {
-          // Fallback to add all workers to the location so some apps (Impala) won't panic.
-          locations.addAll(workerHosts.values());
-          Collections.shuffle(locations);
+          if (!fileBlockInfo.getUfsLocations().isEmpty()) {
+            // Case 1: Fallback to use under file system locations with co-located workers.
+            Map<String, WorkerNetAddress> finalWorkerHosts = workerHosts;
+            locations = fileBlockInfo.getUfsLocations().stream()
+                .map(location -> finalWorkerHosts.get(HostAndPort.fromString(location).getHostText()))
+                .filter(Objects::nonNull).collect(toList());
+          }
```

Comment:
I'm not sure if the performance of this is reasonable if we have a lot of workers. Does `getHostToWorkerMap` end up making copies of the map?

Reply:
it ends up w/ a master rpc and makes a copy

Reply:
this copy is once per file. we can avoid this rpc in the remote ufs case if we embed that info in an additional field in FileStatus

Reply:
Another solution could be to cache the map of eligible workers and have some policy for refreshing the cache? Otherwise we could make a lot of maps which could be of nontrivial size.

Reply:
+1. i'll put in a time based refresh